### PR TITLE
metrics: add scheduling hot peer load

### LINKED ISSUE #5527: metrics: pending influence of hot scheduler is empty
### What did you do?

watch metrics
image

### What did you expect to see?

no empty

### What did you see instead?

empty

### PR DESCRIPTION
**Pull request by lhy1024**
Signed-off-by: lhy1024 <admin@liudos.us>

### What problem does this PR solve?

Issue Number: Fix #5527 

It can be reviewed after #5521 merged

### What is changed and how does it work?

1. collect hot peer dim load in `tryAddPendingInfluence`
2. collect all store pending influence in hot scheduler.
image

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/schedulers/hot_region.go [hunk 1]
```diff
@@ -633,13 +633,15 @@ func (bs *balanceSolver) tryAddPendingInfluence() bool {
...
 	}
+	bs.collectHotPeerLoad(bs.best.mainPeerStat)
 	// revert peers
 	if bs.best.revertPeerStat != nil {
 		infl = statistics.Influence{Loads: make([]float64, statistics.RegionStatCount), Count: 1}
 		bs.rwTy.SetFullLoadRates(infl.Loads, bs.best.revertPeerStat.GetLoads())
 		if !bs.sche.tryAddPendingInfluence(bs.ops[1], dstStoreID, srcStoreID, infl, maxZombieDur) {
 			return false
 		}
+		bs.collectHotPeerLoad(bs.best.revertPeerStat)
```
Comment (HunDunDM) [inline server/schedulers/hot_region.go:644]: Maybe use `infl` to directly collect changes in 6 kinds?

File: server/schedulers/hot_region.go [hunk 2]
```diff
@@ -248,6 +248,16 @@ func (h *hotScheduler) summaryPendingInfluence() {
...
 		}
 	}
+	for storeID, info := range h.stInfos {
+		storeLabel := strconv.FormatUint(storeID, 10)
+		if infl := info.PendingSum; infl != nil {
+			for _, rwTy := range []statistics.RWType{statistics.Read, statistics.Write} {
+				for dim, kind := range rwTy.RegionStats() {
+					hotPendingSum.WithLabelValues(h.GetName(), storeLabel, rwTy.String(), dimToString(dim)).Set(infl.Loads[kind])
+				}
+			}
```
Comment (rleungx) [inline server/schedulers/hot_region.go:258]: almost the same as line 275 to line 279

File: server/schedulers/metrics.go [hunk 3]
```diff
@@ -125,4 +142,6 @@ func init() {
 	prometheus.MustRegister(opInfluenceStatus)
 	prometheus.MustRegister(tolerantResourceStatus)
 	prometheus.MustRegister(hotPendingStatus)
+	prometheus.MustRegister(hotPendingSum)
```
Comment (rleungx) [inline server/schedulers/metrics.go:145]: Do these metrics need to be reset once the leader is transferred?

File: server/schedulers/hot_region.go (hunk 2 above)
Comment (lhy1024) [reply server/schedulers/hot_region.go:258]: line251 collect pending influence for each store, and line275 collect for each scheduling region.

File: server/schedulers/metrics.go (hunk 3 above)
Comment (HunDunDM) [reply server/schedulers/metrics.go:145]: After remove store, we also need to clear the corresponding store information.

**Comment by nolouch** [comment]
ci still failed.

File: pkg/mock/mockcluster/mockcluster.go [hunk 4]
```diff
@@ -778,6 +778,11 @@ func (mc *Cluster) AddSuspectRegions(ids ...uint64) {
 	}
 }
 
+// SetHotPendingInfluenceMetrics mock method
+func (mc *Cluster) SetHotPendingInfluenceMetrics(storeLabel, rwTy, dim string, load float64) {
+
```
Comment (nolouch) [inline pkg/mock/mockcluster/mockcluster.go:783]: remove empty line

File: server/cluster/cluster.go [hunk 5]
```diff
@@ -1797,6 +1797,11 @@ func (c *RaftCluster) deleteStoreLocked(store *core.StoreInfo) error {
 	return nil
 }
 
+// SetHotPendingInfluenceMetrics sets pending influence in hot scheduler.
```
Comment (nolouch) [inline server/cluster/cluster.go:1800]: Why need a function?

File: server/schedulers/hot_region.go [hunk 6]
```diff
@@ -262,6 +270,9 @@ func (h *hotScheduler) tryAddPendingInfluence(op *operator.Operator, srcStore, d
 	h.regionPendings[regionID] = influence
 
 	schedulerStatus.WithLabelValues(h.GetName(), "pending_op_infos").Inc()
+	statistics.ForeachRegionStats(func(rwTy statistics.RWType, dim int, kind statistics.RegionStatKind) {
+		hotPeerHist.WithLabelValues(h.GetName(), rwTy.String(), statistics.DimToString(dim)).Observe(infl.Loads[kind])
```
Comment (nolouch) [inline server/schedulers/hot_region.go:274]: Can we use static metrics?

File: server/cluster/cluster.go (hunk 5 above)
Comment (lhy1024) [reply server/cluster/cluster.go:1800]: If metrics are in `cluster` or `statistic`, we have no way to collect scheduler data; if mertics is in `scheduler`, we have no way to know when the store goes offline or the pd transfer leader. So I put metrics in `cluster` to clear it in time, and use this function to collect `scheduler` statistics.

File: server/schedulers/hot_region.go (hunk 6 above)
Comment (lhy1024) [reply server/schedulers/hot_region.go:274]: The reason is similar with https://github.com/tikv/pd/pull/5525#discussion_r980842651

**Comment by nolouch** [comment]
PTAL @HunDunDM

File: server/schedulers/metrics.go [hunk 7]
```diff
@@ -109,9 +109,18 @@ var hotPendingStatus = prometheus.NewGaugeVec(
...
+		Help:      "Pending influence status in hot region scheduler.",
 	}, []string{"type", "source", "target"})
 
+var hotPeerHist = prometheus.NewHistogramVec(
+	prometheus.HistogramOpts{
+		Namespace: "pd",
+		Subsystem: "scheduler",
+		Name:      "hot_peer",
+		Help:      "Bucketed histogram of the scheduling hot peer.",
+		Buckets:   prometheus.ExponentialBuckets(1, 2, 20),
```
Comment (nolouch) [inline server/schedulers/metrics.go:121]: Bucket may not suitable? such as the read flow may 500MB.

File: server/schedulers/hot_region.go [hunk 8]
```diff
@@ -248,6 +248,14 @@ func (h *hotScheduler) summaryPendingInfluence() {
 			to.AddInfluence(&p.origin, weight)
 		}
 	}
+	for storeID, info := range h.stInfos {
```
Comment (nolouch) [inline server/schedulers/hot_region.go:251]: How to clear the stale metrics? like the store is removing or gone.

File: server/cluster/cluster.go (hunk 5 above)
Comment (nolouch) [reply server/cluster/cluster.go:1800]: > we have no way to know when the store goes offline or the pd transfer leader

I didn't the relative logic to handle offline or transfer leader.

**Comment by nolouch** [comment]
Could show the grafana pannel?

File: server/cluster/cluster.go (hunk 5 above)
Comment (lhy1024) [reply server/cluster/cluster.go:1800]: https://github.com/tikv/pd/blob/e74488f8b121adb751757d071fe3b71993c7b5b9/server/cluster/coordinator.go#L612

File: server/schedulers/hot_region.go (hunk 8 above)
Comment (lhy1024) [reply server/schedulers/hot_region.go:251]: https://github.com/tikv/pd/blob/e74488f8b121adb751757d071fe3b71993c7b5b9/server/cluster/coordinator.go#L606

File: server/schedulers/metrics.go (hunk 7 above)
Comment (nolouch) [reply server/schedulers/metrics.go:121]: ping @lhy1024

**Review by nolouch — APPROVED** [review]

**Review by HunDunDM — APPROVED** [review]

**Pull request merged**